### PR TITLE
added similarity functions

### PR DESCRIPTION
**Pull request by HardiRathod**

### CODE REVIEW

File: tl/features/cell_context_matches.py [hunk 1]
```diff
@@ -185,10 +189,13 @@ def initialize(self, input_df, context_dict, label_column):
                                                                      input_df['kg_aliases']):
             kg_id_context = context_dict.get(kg_id, None)
             kg_labels = []
-            if kg_id_label_str and kg_id_label_str.strip() != "":
-                kg_labels.append(kg_id_label_str.strip())
-            if kg_id_alias_str and kg_id_alias_str.strip() != "":
-                kg_labels.append(kg_id_alias_str.strip())
+
+            if not (isinstance(kg_id_label_str, float)):
```
Comment (saggu) [inline tl/features/cell_context_matches.py:193]: because of line # 176 and 177, this is not required

Comment (HardiRathod) [reply tl/features/cell_context_matches.py]: Removed.

File: tl/features/cell_context_matches.py [hunk 2]
```diff
@@ -213,16 +218,97 @@ def initialize(self, input_df, context_dict, label_column):
                                            score=context_result['score']
                                            )
 
+        property_val_df, important_properties = self.compute_property_scores(row_col_label_dict, n_cntxt_columns, num_rows)
```
Comment (saggu) [inline tl/features/cell_context_matches.py:221]: lines 221 - 247 , put them in a new function

Comment (saggu) [reply tl/features/cell_context_matches.py:221]: does this needs to be in the initialize function?

File: tl/features/cell_context_matches.py [hunk 3]
```diff
@@ -213,16 +218,97 @@ def initialize(self, input_df, context_dict, label_column):
...
+            row_col = r_c.split("_")
+            row_1 = row_col[0]
+            col_1 = row_col[1]
+            for cols in range(n_cntxt_columns):
+                if int(cols) != int(col_1):
+                    m = self.ccm_dict[r_c].get_properties(cols, q_node_dependent=True)
+                    int_prop = pd.DataFrame(m, columns = ["property_", "type", "best_score", "n_occurences"])
+                    int_prop['row'] = int(row_1)
+                    int_prop['column'] = int(col_1)
+                    int_prop['col_2'] = int(cols)
```
Comment (saggu) [inline tl/features/cell_context_matches.py:262]: follow the convention, `col_2` should be `col2`

File: tl/features/cell_context_matches.py [hunk 4]
```diff
@@ -213,16 +218,97 @@ def initialize(self, input_df, context_dict, label_column):
...
+        if " " in to_match_2:
+            split_v = to_match_1.split(" ")
+            for s in split_v:
+                if not s == ".":
+                    new_s = s.replace(".", "0")
+                    if new_s.isnumeric():
+                        num_v = s
+
+        if (to_match_1.isnumeric() or to_match_2.isnumeric() or num_v is not None) and (to_match_1.count(".") <= 1):
+            col2_string_datatype = ["d", "q"]
```
Comment (saggu) [inline tl/features/cell_context_matches.py:297]: we need to discuss this piece of code, the type is coming from context(ES) why do you need to determine type again?

File: tl/features/cell_context_matches.py [hunk 5]
```diff
@@ -213,16 +218,97 @@ def initialize(self, input_df, context_dict, label_column):
...
+        else:
+            col2_string_datatype = ["i"]
+
+        if "i" in col2_string_datatype and "," in col2_string_datatype:
```
Comment (saggu) [inline tl/features/cell_context_matches.py:301]: what is going on here? t

File: tl/features/cell_context_matches.py [hunk 6]
```diff
@@ -233,14 +319,61 @@ def compute_context_similarity(self,
...
+    def computes_string_similarity(self, values: List[str], col2_string_list: List[str], col2_string_datatype: List[str], values_type: str) -> Tuple[float, str]:
+        # Estimate if there's a match between the types.
+        if values_type not in col2_string_datatype:
+            return 0.0, ""
+        max_sim = 0.0
+        best_matched = ""
+        if values_type == "i":
+            for col_split in col2_string_list:
+                for val in values:
+                    sim = similarity.hybrid.symmetric_monge_elkan_similarity(self.preprocess(val),
```
Comment (saggu) [inline tl/features/cell_context_matches.py:349]: i think we discussed a long time ago, that the rltk takes a threshold which should make this code faster, where is that parameter?

**Review by saggu — CHANGES_REQUESTED** [review]
added some comments, needs change

File: tl/features/cell_context_matches.py (hunk 3 above)
Comment (HardiRathod) [reply tl/features/cell_context_matches.py]: corrected.

File: tl/features/cell_context_matches.py (hunk 2 above)
Comment (HardiRathod) [reply tl/features/cell_context_matches.py]: completed.

File: tl/features/cell_context_matches.py (hunk 4 above)
Comment (HardiRathod) [reply tl/features/cell_context_matches.py:299]: This type is not for the context from ES. This is for the given context from the table.

File: tl/features/cell_context_matches.py (hunk 5 above)
Comment (HardiRathod) [reply tl/features/cell_context_matches.py]: This is for separators. Each token split by the separator will be individually tried to match.

File: tl/features/cell_context_matches.py (hunk 6 above)
Comment (HardiRathod) [reply tl/features/cell_context_matches.py:351]: I did not add the parameter for this section as until then I had not started taking threshold parameters in the code.

**Pull request merged**